python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+        activity,
+        *,
+        # Set the name override. If not set, uses the @activity.run decorator
+        # name. If that's not set, uses the function/method name or class name
+        # if it has __call__. If it's a lambda or nested function, a name must
+        # be provided (or it can be decorated of course).
+        name: str = None
+        # ...
+    ) -> None:
+        pass
+
+    # Cannot be called after run started. Typings and separate @overload
+    # signatures omitted for brevity.
+    def register_activities(
+        # This must be an object or list or dictionary. If an object, this only
+        # uses methods with @activity.run decorators. Dictionaries and lists
+        # don't require the decorator. If a dictionary, the key is the name.
+        # This is just a helper and is the equivalent of just calling
+        # register_activity with all applicable functions.
+        activities,
+        # ...
+    ) -> None:
+        pass
+
+    async def run() -> None:
+        pass
+```
+
+Notes:
+
+* We have left off workflow registration until a later phase
+* Stopping the worker is done via cancellation of the run task/future
+* Internally, activities will run in the same event loop `run()` is called on
+* By default will fail already-registered activities
+* Should we accept a set of activities to register in `__init__`?
+* Does exporting `temporalio.Worker` directly cause any packaging issues if we want to separate the client as its own
+  published package?
+
+### Activity Definition/Registration
+
+```python
+from temporalio import activity
+
+# This adds activity details if they are present. Maybe we should have a
+# temporalio.LoggerAdapter that works for activities _and_ workflows?
+my_activity_logger = activity.LoggerAdapter(logging.getLogger(__name__))
+
+
+def simple_activity(arg1, arg2):
+    # Uses contextvars internally, which means it'll fail in another context but
+    # we will provide the ability to set this for testing situations
+    attempts = activity.info().attempts
+    # This logger has an adapter to include custom details but still named after
+    # this module which is a common pattern
+    my_activity_logger.debug("Attempt %s", attempts)
+    # This logger, on the activity module, uses the name of the activity module.
+    # So that's why we offer either the adapter or the logger.
+    activity.logger().debug("Attempt %s", attempts)
+
+
+@activity.run(name="MyActivity")
+async def another_activity(arg1):
+    # Showing heartbeat and cancellation. In Python 3.7 we cannot give a
+    # specific error to Task/Future.cancel (only in 3.9). So activity
+    # cancellations appear like any other asyncio cancellations (which may be a
+    # good thing anyways).
+    try:
+        while True:
+            await asyncio.sleep(1)
+            await activity.heartbeat("some details")
+    except asyncio.CancelledError as exc:
+        raise RuntimeError("someone cancelled me!") from exc
```

Comment:
We will also want something like https://typescript.temporal.io/api/classes/activity.context/#cancelled

(recall Local acts can be cancelled/timed-out w/o heartbeating)

Not having our own type is a biiit rough. B/c you can imagine the user cancelling a task they themselves created inside the activity, and the error thrown there is now indistinguishable from a "Your whole activity got cancelled" error, since AFAICT there's not any extra information on the [type](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError) (not the hugest fan of python docs...) -- and clearly we'll want some way to make that distinction. I suppose we can always just cram in some extra field or whatever since python, but that'd be hard to discover.

Reply:
> We will also want something like https://typescript.temporal.io/api/classes/activity.context/#cancelled

Can do

> Not having our own type is a biiit rough.

We don't in Go, we use the languages native contextual cancellation process. I agree we need our own version, but from above:

>     # Showing heartbeat and cancellation. In Python 3.7 we cannot give a
>     # specific error to Task/Future.cancel (only in 3.9). So activity
>     # cancellations appear like any other asyncio cancellations (which may be a
>     # good thing anyways).

I admittedly have not done enough research here, but I agree if I can use our own cancellation error I will. But I was just worried reading docs that issuing a cancel in asyncio world in the version of Python we're targeting doesn't allow you to customize the type.

Reply:
> Which raises another complication with non async activity functions.

They can't heartbeat anyways, heh. We should "accept" non-async activities since they can be just quick CPU bound tasks, but in all our examples, we should `async def` activities just to make it clear they are the suggested approach.

Reply:
I have added a section on activity utilities that provides the cancellation options.

Reply:
The thing I'm still curious about here is if we can attach discoverable information to `asyncio.CancelledError`? It doesn't seem like we really can. The added ability to get the cancel info helps for sure but if we could it'd be nice to have it on the exception itself somehow

Reply:
The answer to that will require experimentation. Task factories and event loop inheritance has been mentioned as possible options.